### ISSUE TITLE
cannot define richcmp special functions in cdef class

### ISSUE THREAD

**Issue by rwst** [cython/cython#2019]
Cython-0.27.2. When I try to add an `__eq__` member like this:
```
    def __eq__(self, right):
        print("co", self, right)
        return True
```
at this codepoint: https://github.com/sagemath/sage/blob/master/src/sage/libs/pynac/constant.pyx#L99
I get a compile error:
```
[sagelib-8.1.rc2] [1/2] Cythonizing sage/libs/pynac/constant.pyx
[sagelib-8.1.rc2] [2/2] Cythonizing sage/libs/pynac/pynac.pyx
[sagelib-8.1.rc2] Traceback (most recent call last):
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Build/Dependencies.py", line 1179, in cythonize_one_helper
[sagelib-8.1.rc2]     return cythonize_one(*m)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Build/Dependencies.py", line 1144, in cythonize_one
[sagelib-8.1.rc2]     result = compile([pyx_file], options)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/Main.py", line 695, in compile
[sagelib-8.1.rc2]     return compile_multiple(source, options)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/Main.py", line 673, in compile_multiple
[sagelib-8.1.rc2]     result = run_pipeline(source, options, context=context)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/Main.py", line 494, in run_pipeline
[sagelib-8.1.rc2]     err, enddata = Pipeline.run_pipeline(pipeline, source)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/Pipeline.py", line 354, in run_pipeline
[sagelib-8.1.rc2]     data = run(phase, data)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/Pipeline.py", line 334, in run
[sagelib-8.1.rc2]     return phase(data)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/Pipeline.py", line 52, in generate_pyx_code_stage
[sagelib-8.1.rc2]     module_node.process_implementation(options, result)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/ModuleNode.py", line 141, in process_implementation
[sagelib-8.1.rc2]     self.generate_c_code(env, options, result)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/ModuleNode.py", line 380, in generate_c_code
[sagelib-8.1.rc2]     self.generate_module_init_func(modules[:-1], env, globalstate['init_module'])
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/ModuleNode.py", line 2377, in generate_module_init_func
[sagelib-8.1.rc2]     self.generate_type_init_code(env, code)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/ModuleNode.py", line 2864, in generate_type_init_code
[sagelib-8.1.rc2]     self.generate_type_ready_code(env, entry, code)
[sagelib-8.1.rc2]   File "/home/ralf/sage/local/lib/python2.7/site-packages/Cython/Compiler/ModuleNode.py", line 2963, in generate_type_ready_code
[sagelib-8.1.rc2]     slot = TypeSlots.method_name_to_slot[func.name]
[sagelib-8.1.rc2] KeyError: u'__eq__'
```
This is the same with `lt`, `ne` and so on...

**Comment by rwst** [comment]
Let me add that I need this to fix https://trac.sagemath.org/ticket/24255 which is a bug exactly because of the implemented odd behaviour mentioned in #1848

**Comment by scoder** [comment]
There was a problem with special comparison methods that had docstrings. Either remove the docstring as a quick-fix, or wait for Cython 0.28.

**Comment by hearot** [comment]
Just had a problem with this today.

**Comment by scoder** [comment]
@hearot: was this with 0.28.x, or which Cython version are you using? Since the bug described here is already fixed in 0.28, could you please open a new ticket with all necessary information if you run into a problem with 0.28.3?

**Comment by hearot** [comment]
@scoder Sure.
